crd-generator/collector/src/main/java/io/fabric8/crd/generator/collector/JandexCustomResourceClassScanner.java

```diff
@@ -0,0 +1,150 @@
+ * existing Jandex index if found in the source. Otherwise, the index will be created on the fly.
+ * </p>
+ *
+ * @see Index
+ * @see JandexIndexer
+ * @see JandexUtils
+ */
+public class JandexCustomResourceClassScanner {
+
+  private final List<IndexView> indices = new LinkedList<>();
+  private final Set<File> filesToIndex = new HashSet<>();
+
+  /**
+   * If <code>true</code>, indices will always be created even if an index exists at the source.
+   */
+  private boolean forceIndex = false;
+
+  public JandexCustomResourceClassScanner withForceIndex(boolean forceIndex) {
+    this.forceIndex = forceIndex;
+    return this;
+  }
+
+  public JandexCustomResourceClassScanner withIndex(IndexView... index) {
+    if (index != null) {
+      withIndices(Arrays.asList(index));
+    }
+    return this;
+  }
+
+  public JandexCustomResourceClassScanner withIndices(Collection<IndexView> indices) {
+    if (indices != null) {
+      indices.stream()
+          .filter(Objects::nonNull)
+          .forEach(this.indices::add);
+    }
+    return this;
+  }
+
+  public JandexCustomResourceClassScanner withFileToIndex(File... files) {
+    if (files != null) {
+      withFilesToIndex(Arrays.asList(files));
+    }
+    return this;
+  }
+
+  public JandexCustomResourceClassScanner withFilesToIndex(Collection<File> files) {
+    if (files != null) {
+      files.stream()
+          .filter(Objects::nonNull)
+          .forEach(this.filesToIndex::add);
+    }
+    return this;
+  }
+
+  /**
+   * Finds Custom Resource classes in the pre-configured scope.
+   *
+   * @return the Custom Resource classes
+   */
+  public Collection<String> findCustomResourceClasses() {
+    List<IndexView> actualIndices = new ArrayList<>(indices);
+    actualIndices.add(createBaseIndex());
+
+    IndexView index = JandexUtils.createIndex(actualIndices, filesToIndex, forceIndex);
+
+    return findCustomResourceClasses(index).stream()
+        .map(ClassInfo::toString)
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Finds Custom Resource classes in an index.
+   *
+   * @param index the index
+   * @return the Custom Resource classes
+   *
+   * @see JandexCustomResourceClassScanner#createBaseIndex()
+   */
+  private List<ClassInfo> findCustomResourceClasses(IndexView index) {
+    // Only works if HasMetadata and all intermediate classes (like CustomResource)
+    // are included in given index.
+    return index.getAllKnownImplementors(HasMetadata.class)
+        .stream()
+        .filter(classInfo -> classInfo.hasAnnotation(Group.class))
+        .filter(classInfo -> classInfo.hasAnnotation(Version.class))
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Creates the base index required to scan for Custom Resource classes.
+   *
+   * @return the base index.
+   *
+   * @see JandexCustomResourceClassScanner#findCustomResourceClasses(IndexView)
+   */
+  private Index createBaseIndex() {
```

Comment:
This can be optimized if required. 
Existing indices of the fabric8 modules can be used or a specialized index can be created at build time.

Do we need further optimization within this PR?

Reply:
We probably don't need optimization here, but I agree that we should reuse the existing index is possible - since users wil need to modify their pom / gradle file to use these plugins, it doesn't seem out of the question to just document also requiring the jandex plugin be configured as well for projects that want to create crds.

Reply:
I'm currently writing some usage docs about the different use cases, including when additional configuration is required. This should help to get an overview from the user perspective.

Some notes on a more technically perspective:

- Jandex is only required to "scan" for Custom Resource classes. If the list of classes is given, no scanning at all is required. (The classloader does the heavy lifting)
- The implementation detects existing indices in the scan targets automatically and prefers those. Only if a scan target does not contain an index, an own in-memory index will be created. This means having an existing index is just optimization, never a requirement.
- Custom Resource classes must be loaded before they can be used by the CRDGenerator itself. Any dependency which is required to load those classes must be in the classpath.
- A scan target must be also in the classpath. But not every classpath element is a scan target.
- Additional indexed classes beside the "base-index" are only required, if the user wants to scan a target with Custom Resource classes which don't implement the CustomResource or HasMetadata interface directly.  In this case, the user must ensure that the used intermediate class is included somehow in the underlying composite index. For example, he can add the dependency which includes the class to the list of scan targets, so that an index is automatically created if needed.

In conclusion, a user doesn't need to know anything about Jandex. Even for the edge case from the last bullet point, the user only has to care about including all "relevant" classes to the list of scan targets and not Jandex specifically.

(But if his project already uses Jandex, the scans will be a little bit faster.)

Keeping this in mind, optimizing the creation of the base index is mostly about performance.
Only if there are other "intermediate" classes like "CustomResource", then we should add them.

Reply:
I wrote an integration test for the "intermediate class" case (last bullet point from above):

https://github.com/fabric8io/kubernetes-client/pull/5979/commits/ddc3bec8dda01d0ee053d44c6240bb7840f6552d#diff-0488fc4e2f91d3e88ce94d75d1317631a14f2854194a2cf68354be4712e9c2cb

In this case, the user must add the dependency, which contains the intermediate class, to the list of dependencies to scan. Under the hood the classes are now added to the composite index and Custom Resources Classes will be found by the collector.

Reply:
Are any questions on this open or can we resolve this conversion?